prov/gni/src/gnix_rma.c

```diff
@@ -399,50 +745,81 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
 		return -FI_EINVAL;
 	}
 
-	rdma = len >= ep->domain->params.rma_rdma_thresh;
-
-	/* need a memory descriptor for all RDMA and reads */
-	if (!mdesc && (rdma || fr_type == GNIX_FAB_RQ_RDMA_READ)) {
-		rc = gnix_mr_reg(&ep->domain->domain_fid.fid, (void *)loc_addr,
-				 len, FI_READ | FI_WRITE, 0, 0, 0, &auto_mr,
-				 NULL);
-		if (rc != FI_SUCCESS) {
-			GNIX_INFO(FI_LOG_EP_DATA,
-				  "Failed to auto-register local buffer: %d\n",
-				  rc);
-			return rc;
-		}
-		flags |= FI_LOCAL_MR;
-		mdesc = (void *)auto_mr;
-		GNIX_INFO(FI_LOG_EP_DATA, "auto-reg MR: %p\n", auto_mr);
-	}
-
 	/* find VC for target */
 	rc = _gnix_ep_get_vc(ep, dest_addr, &vc);
 	if (rc) {
 		GNIX_INFO(FI_LOG_EP_DATA,
 			  "_gnix_ep_get_vc() failed, addr: %lx, rc:\n",
 			  dest_addr, rc);
-		goto err_get_vc;
+		return rc;
 	}
 
 	/* setup fabric request */
 	req = _gnix_fr_alloc(ep);
 	if (!req) {
 		GNIX_INFO(FI_LOG_EP_DATA, "_gnix_fr_alloc() failed\n");
-		rc = -FI_ENOSPC;
-		goto err_fr_alloc;
+		return -FI_ENOSPC;
 	}
 
+	rdma = len >= ep->domain->params.rma_rdma_thresh;
+
 	req->type = fr_type;
 	req->gnix_ep = ep;
 	req->vc = vc;
 	req->user_context = context;
 	req->work_fn = _gnix_rma_post_req;
 
-	if (mdesc) {
-		md = container_of(mdesc, struct gnix_fid_mem_desc, mr_fid);
+	if (fr_type == GNIX_FAB_RQ_RDMA_READ &&
+	    (rem_addr & GNI_READ_ALIGN_MASK || len & GNI_READ_ALIGN_MASK)) {
+		/* Copy unaligned data through the inject buffer. */
+		align_buf = req->inject_buf;
+		rc = gnix_mr_reg(&ep->domain->domain_fid.fid, align_buf,
+				GNIX_INJECT_SIZE, FI_READ, 0, 0, 0, &align_mr,
+				NULL);
+		if (rc != FI_SUCCESS) {
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "Failed to auto-register align buffer: %d\n",
+				  rc);
+			goto err_unalign_reg;
+		}
+
+		if (len >= GNIX_RMA_UREAD_CHAINED_THRESH) {
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "Using CT for unaligned GET, req: %p\n",
+				  req);
+			flags |= GNIX_RMA_CHAINED;
+		} else {
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "Using tmp buf for unaligned GET, req: %p\n",
+				  req);
+			flags |= GNIX_RMA_INDIRECT;
+		}
+
+		if (rdma)
+			req->work_fn = _gnix_rma_post_rdma_chain_req;
+
+		GNIX_INFO(FI_LOG_EP_DATA, "align MR: %p\n", align_mr);
 	}
+
+	if (!(flags & GNIX_RMA_INDIRECT) && !mdesc &&
+	    (rdma || fr_type == GNIX_FAB_RQ_RDMA_READ)) {
+		/* We need to auto-register the source buffer. */
+		rc = gnix_mr_reg(&ep->domain->domain_fid.fid, (void *)loc_addr,
+				 len, FI_READ | FI_WRITE, 0, 0, 0, &auto_mr,
+				 NULL);
+		if (rc != FI_SUCCESS) {
```

Comment:
this is okay for now, but eventually I think we want to do something more efficient than registering these small inject buffers.  I could see how for a problematic code we could get 1000s of these small registrations.  @e-harvey is working on a buddy allocator which could eventually be used at this point.